### ISSUE TITLE
Add an "auto" tile mode

### ISSUE THREAD

**Issue by dwbuiten** [xiph/rav1e#1230]
Now that tile support is merged, it would be very useful for rav1e to be able to set a "good" amount of tiles for a given input resolution instead of every downstream user manually setting it.

Not sure if much testing has been done to determine what such a "good" number would be for any given res.

**Comment by tdaede** [comment]
Yeah, I mentioned this on IRC first. Probably the first step would just be a "# of tiles" API that sets horiz and vert tiling appropriately.

**Comment by dwbuiten** [comment]
As in an API that takes a number of tiles and verifies it can log2() it, and sets it? That's easy enough.

As for 'auto' mode, is there any evidence for a minimum tile size at which things get Bad™?

**Comment by tdaede** [comment]
It could actually take the total number of tiles and find the best horizontal + vertical split.

For starters, the maximum tile size is 4kx2k, so we *need* to split before that point. Beyond that, it requires some experimentation I think. It's super application dependent - for high latency streaming I suspect that 2 tiles will be the most you want.

**Comment by dwbuiten** [comment]
Is that 2 tiles total or two tiles per direction?

**Comment by tdaede** [comment]
Total. Basically you want to use as few as possible and segment thread (or frame thread once we get it) as much as possible first. Maybe 4 for 4k, I dunno. They do help with software decoding as well, might be worth benchmarking dav1d on some common platforms where more tiles might make the difference between real time decoding and not.

Now, if you are doing live encoding, this totally changes and you want to do lots of tiles, because segment threading and frame threading aren't usable and the loss of tiles is much lower than the loss of not using all your cores.

**Comment by dwbuiten** [comment]
Hmm, OK. This somewhat negates the usefulness of an 'auto' mode, aside from setting tiles due to max tile size, unless you think it's useful to figure out when two tiles are useful automatically.

**Comment by rzumer** [comment]
I think number of tiles depends largely on use case, but maybe we can devise some formula based on resolution, latency and speed level.

**Comment by tdaede** [comment]
Yeah, in particular I can think of two main use cases right now.

1. Segmented encoding. You want minimum number of tiles. (--tiles 1)

2. Encoding where you're the only thing on the whole machine. (--tiles numcpus)

If you're encoding e.g. just one file locally, you probably actually want (2), as it lets you use a lower speed level, which is a bigger gain than the loss of using tiles, probably. So I was thinking of (2) being the default.

**Comment by tdaede** [comment]
(when we add frame threading, then we should probably prefer that and scale back tile threads appropriately, but until then this seems reasonable)

**Comment by vibhoothiiaanand** [comment]
WIP PR: https://github.com/xiph/rav1e/pull/1370
